core/amber/src/main/scala/edu/uci/ics/amber/engine/common/storage/FileDocument.scala

```diff
@@ -0,0 +1,111 @@
+package edu.uci.ics.amber.engine.common.storage
+import org.apache.commons.vfs2.{FileObject, VFS}
+
+import java.io.{InputStream, OutputStreamWriter}
+import java.net.URI
+import java.util.concurrent.locks.ReentrantReadWriteLock
+
+/**
+  * FileDocument provides methods to read/write a file located on filesystem.
+  * All methods are THREAD-SAFE implemented using read-write lock:
+  * - 1 writer at a time: only 1 thread of current JVM can acquire the write lock
+  * - n reader at a time: multiple threads of current JVM can acquire the read lock
+  * @param uri the identifier of the file. If file doesn't physically exist, FileDocument will create the file during the constructing phase.
+  */
+class FileDocument(val uri: URI) extends VirtualDocument[String] {
```

Comment:
FileDocument should handle types other than strings. I suggest we use FileDocument[T] extends VirtualDocument[T] and do the serialization using kryo. You can refer to what I did in `SequentialRecordStorage`.